core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,530 @@
+            throw new RuntimeException(failure);
+        }
+
+        if (closed) {
+            return;
+        }
+
+        if (isFinished()) {
+            return;
+        }
+
+        if (!blocked.isDone()) {
+            return;
+        }
+
+        if (taskSource == null) {
+            Map<PlanFragmentId, ListenableFuture<List<ExchangeSourceHandle>>> exchangeSourceHandleMap = sourceExchanges.entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getKey, entry -> toListenableFuture(entry.getValue().getSourceHandles())));
+
+            List<ListenableFuture<List<ExchangeSourceHandle>>> blockedFutures = exchangeSourceHandleMap.values().stream()
+                    .filter(future -> !future.isDone())
+                    .collect(toImmutableList());
+
+            if (!blockedFutures.isEmpty()) {
+                blocked = asVoid(allAsList(blockedFutures));
+                return;
+            }
+
+            Multimap<PlanFragmentId, ExchangeSourceHandle> exchangeSources = exchangeSourceHandleMap.entrySet().stream()
+                    .collect(flatteningToImmutableListMultimap(Map.Entry::getKey, entry -> getFutureValue(entry.getValue()).stream()));
+
+            taskSource = taskSourceFactory.create(
+                    session,
+                    stage.getFragment(),
+                    sourceExchanges,
+                    exchangeSources,
+                    stage::recordGetSplitTime,
+                    sourceBucketToPartitionMap,
+                    sourceBucketNodeMap);
+        }
+
+        while (!queuedPartitions.isEmpty() || !taskSource.isFinished()) {
+            while (queuedPartitions.isEmpty() && !taskSource.isFinished()) {
+                List<TaskDescriptor> tasks = taskSource.getMoreTasks();
+                for (TaskDescriptor task : tasks) {
+                    queuedPartitions.add(task.getPartitionId());
+                    partitionToTaskDescriptorMap.put(task.getPartitionId(), task);
+                    sinkExchange.ifPresent(exchange -> {
+                        ExchangeSinkHandle exchangeSinkHandle = exchange.addSink(task.getPartitionId());
+                        partitionToExchangeSinkHandleMap.put(task.getPartitionId(), exchangeSinkHandle);
+                    });
+                }
+                if (taskSource.isFinished()) {
+                    sinkExchange.ifPresent(Exchange::noMoreSinks);
+                }
+            }
+
+            if (queuedPartitions.isEmpty()) {
+                break;
+            }
+
+            int partition = queuedPartitions.peek();
+            TaskDescriptor taskDescriptor = requireNonNull(partitionToTaskDescriptorMap.get(partition), () -> "task descriptor missing for partition: %s" + partition);
+
+            if (acquireNodeFuture == null) {
+                acquireNodeFuture = nodeAllocator.acquire(taskDescriptor.getNodeRequirements());
+            }
+            if (!acquireNodeFuture.isDone()) {
```

Comment:
Should we keep a number of futures (e.g a map requirements->future). There is a chance that we cannot allocate a node for given set of requirements but we would be fine for other set.